### ISSUE TITLE
Security control for user-defined groovy scripts

### ISSUE THREAD

**Issue by alexbudarov** [cuba-platform/cuba#912]
### Description of the bug or enhancement
CUBA and CUBA-based apps use user-defined groovy scripts a lot.
By "user-defined" I mean: input through web UI by user who does not necessarily has administration access to server hosting.

Examples of functionality:
- report bands
- report parameter validation scripts
- security constraints

The problem with groovy is that it can execute arbitrary code, including damaging one, e.g. : 
> System.exit(0); return true

Actually, at the moment you can only give access to edit groovy scripts to users who are either server administrators or other trusted employees.

User-defined groovy scripts need to be constrainted for what java methods and classes they can call.

5-minute googling has shown that there is support for "security sandbox" in groovy:
http://groovy-sandbox.kohsuke.org/
http://melix.github.io/blog/2015/03/sandboxing.html

Looks like we need to define a CUBA API which by default whitelists only safe java APIs.
And allows application developer to programmatically add more trusted classes / methods (e.g. static helper classes or spring beans aimed to reduce script verbosity).

**Comment by jreznot** [comment]
You know you can still cause OOM easily.

**Comment by alexbudarov** [comment]
yes, there are also for example infinite cycles. They are mentioned by link above.

**Comment by alexbudarov** [comment]
Denial-of-service is lesser problem.
The worst thing is ability to execute programs / scripts on server.

**Comment by mariodavid** [comment]
I've looked into this topic a few years ago actually during the implementation of the runtime-diagnose plugin (https://github.com/mariodavid/cuba-component-runtime-diagnose).

As you see I was able to do (more or less) proper security checking for SQL, but not for Groovy. The reason was, that in the blog post you mentioned (http://melix.github.io/blog/2015/03/sandboxing.html), it clearly says: 

> The current list of customizers available in the Groovy distribution, and the currently available sandboxing projects in the wild, are not sufficient to guarantee security of execution of scripts in the general case (dependending, of course, on your users and where the scripts come from).

as well as:

> Last but not least, the solution that I described here is incomplete. 

Situation has not really changed since 2015 when the blog post was published.
The only really secure solution to this problem is to use a whitelist. But this is practically impossible to do if you don't have something like a DSL that should get executed. But this is oftentimes against the idea of the scripts that should get executed in the runtime-diagnose case in particular but in a lot of CUBA cases in general.

Therefore I decided (for my plugins) to instead of controlling the content of the scripts itself, it is more practical to control the ability for users to define those scripts. Like - making sure that only the right people are allowed to actually define the groovy scripts.

Another even more restrictive way would be to instead of letting the users directly create the groovy code, they are only allowed to select a script from a list (that is stored in the DB e.g.) and use is (e.g. in the case of the AppFolders or something). The list of the scripts would only be managed by a subset of those users (e.g. only the admins or something), so a proper review can take place. But this would require you to rework a lot of build it UIs to instead of directly enter source code set a reference to a groovy script.

Bye
Mario
